supervisor/supervisor.go

```diff
@@ -0,0 +1,41 @@
+package supervisor
+
+import (
+	"context"
+	"github.com/open-telemetry/opamp-go/supervisor/types"
+)
+
+// Supervisor allows the management of an agent. Specific agents like an
+// OpenTelemetry collector can implement a custom supervisor that understands
+// the internal workings of the collector. Alternatively, a default supervisor
+// allows for the supervision of arbitrary daemon-like processes and provides
+// management options like remote configuration change, agent upgrades,
+// watchdog capabilities, telemetry reporting, etc.
+//
+// The supervisor wraps the client part of the OpAMP protocol and therefore
+// requires an OpAMP server to be running at some endpoint. There is a 1:1
+// mapping between a supervisor and the OpAMP client. A single supervisor
+// executable, however, can run several supervisors within it thereby managing
+// several distinct agent types.
+//
+// See the supervisor sample configuration file and the associated
+// `types.Configuration` for details.
+type Supervisor interface {
+	// Start the supervisor and begin monitoring the associated agent as
+	// specified by the configuration.
+	//
+	// Start MUST return immediately (potentially with an error if the specified
+	// configuration is incorrect or it fails to parse).
+	//
+	// Start, in most cases, should spawn a Go routine to monitor the agent
+	// under supervision.
+	//
+	// Start, on a given instance, must be called exactly once. Subsequent
+	// calls have no effect.
+	Start(config *types.Configuration) error
```

Comment:
This requires a very specific way of specifying the Configuration. I am not sure what the purpose of such `Supervisor` interface is.

I was imagining a different approach when supervising capabilities are implemented as composable features which can be used to create a generic or specific Supervisor. This requires reversing the interfaces and have e.g. `SupervisableAgent` interface which can be passed to the Supervisor implementation to be managed.

Reply:
This was bit of an oversight on my side. I think I was imagining that we implement the entire supervisor every time we have a new agent that requires supervision. But, many supervising capabilities might be overlapping and hence need not be re-implemented. Let me rethink this a bit and provide an alternate implementation.